### PR TITLE
Implemented module version validation

### LINKED ISSUE #674: Version & Module Validation
# For New Contributors
[How to contribute](https://github.com/Bannerlord-Coop-Team/BannerlordCoop/wiki/Getting-Started-as-a-Contributor)

## Description

Validate that the game, mod & modules are on the same version as the server.

## Intended Design

Validate that the game, mod & modules are on the same version as the server.

## Definition of Done
- [ ] Class level comments exist for all new classes.
- [ ] XUnit tests exist for every method that does not require the game to be ran.

### PR DESCRIPTION
**Pull request by dennispost99**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] All new classes have class-level documentation comments, if there are any at all
- [x] Tests for the changes have been added (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation update
- [ ] Other... Please describe:

## What is the current behaviour?

Modules are not validated

## What is the new behaviour?

Resolves #674 

Modules get validated now. Client modules must match the server modules

## Other information
I have tried to use as much existing code as possible, if something should be outsourced, please write a comment, I will try to change it in time

### CODE REVIEW

File: source/Coop.Core/Client/States/ValidateModuleState.cs [hunk 1]
```diff
@@ -64,6 +67,19 @@ internal void Handle_NetworkClientValidated(MessagePayload<NetworkClientValidate
...
+    internal void Handle_NetworkModuleVersionsValidated(MessagePayload<NetworkModuleVersionsValidated> obj)
+    {
+        if (obj.What.Matches)
+        {
+            network.SendAll(new NetworkClientValidate(controllerIdProvider.ControllerId));
+        }
+        else
+        {
+            messageBroker.Publish(this, new SendInformationMessage(obj.What.Reason));
+            Disconnect();
```
Comment (garrettluskey) [inline source/Coop.Core/Client/States/ValidateModuleState.cs]: Does disconnect properly change the state back to MainMenuState?

File: source/Coop.Core/Server/Connections/Messages/NetworkModuleVersionsValidate.cs [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+/// <summary>
+/// Message from Client to Server for validating the module versions.
+/// Responsibilities
+/// 1. Make sure that all active modules have the same version as on the server
+/// </summary>
+[ProtoContract(SkipConstructor = true)]
+public record NetworkModuleVersionsValidate : ICommand
+{
+    [ProtoMember(1)]
+    public List<NetworkModuleInfo> Modules { get; }
```
Comment (garrettluskey) [inline source/Coop.Core/Server/Connections/Messages/NetworkModuleVersionsValidate.cs]: Change this to an array as it is easier to serialize and takes less bandwidth (kind of pedantic but I noticed this)

File: source/GameInterface/Services/Modules/Validators/ModuleValidator.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace GameInterface.Services.Modules.Validators;
+
+public class ModuleValidator : IModuleValidator
```
Comment (garrettluskey) [inline source/GameInterface/Services/Modules/Validators/ModuleValidator.cs]: I found it a lot easier to have the interface and implementation in the same file,

I suggest moving the IModuleValidator interface definition to this file. Also could you add
```CS
/// <inheritdoc href="IModuleValidator">
```
And also add a summary comment for the IModuleValidator interface

**Comment by garrettluskey** [comment]
Please fix tests so they run on the pipeline

**Review by garrettluskey — CHANGES_REQUESTED** [review]

File: source/Coop.Core/Client/States/ValidateModuleState.cs (hunk 1 above)
Comment (dennispost99) [reply source/Coop.Core/Client/States/ValidateModuleState.cs]: The status is set directly now

File: source/GameInterface/Services/Modules/Validators/ModuleValidator.cs (hunk 3 above)
Comment (dennispost99) [reply source/GameInterface/Services/Modules/Validators/ModuleValidator.cs]: I adapted it to the code conventions of the project

**Review by garrettluskey — DISMISSED** [review]

**Review by garrettluskey — APPROVED** [review]

**Pull request merged**